### DISCUSSION TITLE
Improvement of the static code analysis

### DISCUSSION THREAD

**Discussion by MessiasLima** [patrykandpatrick/vico#542] [Ideas]
# Problem
In the current build logic, ktlint is downloaded on Github pipeline and the verifications happens there. That makes the build logic lightweight but it adds more steps in the pull request review.

In fact, the maintainer has to give permission to run the pipeline just because there is a blank line in the wrong place.

# Solution
Set the static code verification to run locally, before the user pushes the committed changes. That will add one more automatic step before the actual push but it will save lots of time from github CI, from the contributor and from the maintainer.

## How?
That can be achieved with 2 gradle plugins: 
- detekt (io.gitlab.arturbosch.detekt): That will do the code checks, including the ktlint style verifications
- git hook (com.star-zero.gradle.githook): That will set up a git hook to run the detekt checks before every push

## The effort required
The effort to do that is not very high. I can do it and open a pull request. I just wanted to ask for the opinion of yours before start working on it.

**Comment by patrickmichalik** [comment]
Hello! Thanks for your interest in contributing. I’d say it’s relatively common to have a GitHub Actions workflow that runs a code formatter. Such a solution and local formatting checks aren’t mutually exclusive. While Ktlint isn’t included in the Vico project, you can download it and run it on your local machine before pushing your changes. There’s no need to wait for it to run on GitHub Actions. We’ve added a note about Ktlint [here](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.6/contributing/#pull-requests).

That said, a Gradle plugin for Ktlint could certainly be added. Ktlint then wouldn’t have to be downloaded separately. We used to use Detekt, but we had a few gripes with it. How about Spotless (with only Ktlint)?

A hook sounds good too, but we’d like to avoid running configuration changes automatically. Our suggestion would be to add a Bash script or a Gradle task that you can run to add the hook. Since hook configuration is simple, it would be preferable to avoid adding a dependency just for this.

We’d be happy to accept a pull request for all or part of the above. If you have any questions, feel free to let me know. Cheers!
